### PR TITLE
Bug 1200091 - PBAP reference design.

### PR DESCRIPTION
**Pull request by weilonge**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2664cb4c1539417f13986ea944792a27d84b4e3e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8ef13cc5f52f6edd0afda62981aad6a1df3a6c24

File: apps/communications/pbap/js/call_log.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+/* global applications, BrowserConfigHelper, AppWindow */
+/* jshint nonew: false */
+
+(function(exports) {
+  function PbapCallLog() {
+    this.preHandleEvent = this.preHandleEvent.bind(this);
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/call_log.js]: This isn't doing anything, why do you need this? If there's no specific need then this file must be removed.

File: apps/communications/pbap/js/pbap.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+(function (exports){
+  var managerBt = navigator.mozBluetooth;
+  var managerContacts = navigator.mozContacts;
+  var adapter;
+  var pb;
+
+  function pbapInit(){
+    pb = new PbapPhonebook();
+    if (!managerBt) {
+      console.log('PBAP', 'No Bluetooth Manager');
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: Make this console.error()

File: apps/communications/pbap/js/pbap.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      return;
+    }
+    var req = managerBt.getDefaultAdapter();
+    req.onsuccess = function bt_getAdapterSuccess() {
+      console.log('PBAP', 'adapter success');
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: Do not log a message if we're following the normal flow

File: apps/communications/pbap/js/pbap.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      adapter = req.result;
+      registerPbapEvent();
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: There is no need to have a separate method for this, just inline the code here, also please use addEventListener() to register events, it's preferred

File: apps/communications/pbap/js/pbap.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  function registerPbapEvent(){
+    adapter.onpullphonebookreq = pullphonebook;
+    adapter.onpullvcardentryreq = pullvcardentry;
+    adapter.onpullvcardlistingreq = pullvcardlisting;
+  }
+
+  LazyLoader.load([
+                   '/shared/js/text_normalizer.js',
+                   '/shared/js/contact2vcard.js',
+                   '/shared/js/setImmediate.js'
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: You do not seem to be using setImmediate(), please remove this source file

File: apps/communications/pbap/js/phonebook.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
+'use strict';
+/* global applications, BrowserConfigHelper, AppWindow */
+/* jshint nonew: false */
+
+(function(exports) {
+  function PbapPhonebook() {
+    this.ctsMan = navigator.mozContacts;
+  }
+
+  function generateVcardList(cacheList){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit leave a space between the closing parentheses and the opening brace

File: apps/communications/pbap/js/phonebook.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      function appendOwnerVcard () {
+        return {
+          name: 'Owner;Vcard',
+          ctsId: 'ownerID',
+        };
+      }
+
+      return new Promise((resolve, reject) => {
+        this._listingCache = [];
+        //this._listingCache.push(appendOwnerVcard());
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: Remove this commented out code

File: apps/communications/pbap/js/phonebook.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      var content = XML_HEADER + lines.join('\n') + '\n' + XML_FOOTER;
+      resolve({
+        xml: content,
+        size: cacheList.length
+      });
+    });
+  }
+
+  PbapPhonebook.prototype = {
+    pullVcardListing: function (request){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit: leave a space between the closing parentheses and the opening brace

File: apps/communications/pbap/js/phonebook.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        contacts.forEach((item) => {
+          var cacheItem = {};
+          cacheItem.name = item.familyName + ';' + item.givenName;
+          cacheItem.ctsId = item.id;
+          this._listingCache.push(cacheItem);
+        });
+        resolve(this._listingCache);
+      });
+    },
+    pullVcardEntry: function (request){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit: leave a space between the closing parentheses and the opening brace

File: apps/communications/pbap/js/phonebook.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        request.onsuccess = function () {
+          resolve(request.result);
+        };
+        request.onerror = function (error) {
+          console.error('PBAP', 'Phonebook error: ', error);
+          reject(error);
+        };
+      });
+    },
+    filterContacts: function (filters){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit: leave a space between the closing parentheses and the opening brace

File: apps/communications/pbap/js/phonebook.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+          resolve(request.result);
+        };
+        request.onerror = function (error) {
+          console.error('PBAP', 'Phonebook error: ', error);
+          reject(error);
+        };
+      });
+    },
+
+    sortAllContacts: function (filters){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit: leave a space between the closing parentheses and the opening brace

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=88270e063f70518241d28e8fb01d4d331dd5f36b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a37ea9820be358a44521bac6d9feaad1f262192a

File: apps/communications/pbap/js/pbap.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+/*
+    Phone Book Access Profile (PBAP) is a profile that enables devcices to
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: nit: Please keep the indentation of comments consistent and use jsdoc comments, for example here I'd prefer
```
/**
 * Phone Book Access Profile (PBAP) is a profile that enables devices to
 * exchange phonebook objects via Bluetooth. After a Bluetooth connection is
 * established, ...
 */
```

Likewise for the other comments.

File: apps/communications/pbap/js/pbap.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  from the client to response the phonebook records or list, and all three kinds
+  of request from the client are:
+  * pullphonebook - pull all phonebook records based on the constraint of the
+    filter object.
+  * pullvcardentry - pull a specific contacts.
+  * pullvcardlisting - pull a XML format list of all contacts which filtered by
+    a filter argument.
+*/
+
+(function (exports){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: nit: Please leave a space between the closing parentheses and opening brace.

File: apps/communications/pbap/js/pbap.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+  var managerBt = navigator.mozBluetooth;
+  var managerContacts = navigator.mozContacts;
+  var adapter;
+  var pb;
+
+  function pbapInit(){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/pbap.js]: nit: Likewise

File: apps/communications/pbap/js/phonebook.js [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        Sort and filter the records based on the filter provided by BT profile.
+      In this prototype, all contacts will be retrieved.
+    */
+    sortAllContacts: function (filters) {
+      return new Promise((resolve, reject) => {
+        filters = filters || {};
+        var contacts = [];
+        var request = this.ctsMan.getAll(filters);
+        request.onsuccess = function () {
+          if(this.result){
```
Comment (gabrielesvelto) [inline apps/communications/pbap/js/phonebook.js]: nit: Please leave a space between the if and the opening brace and between the closing parentheses and opening brace.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=852df8fbafa70ed855cb0b4b9cf116e2d14effa5

File: apps/communications/pbap/js/pbap.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
+'use strict';
```
Comment (pasawaytilldie23) [inline apps/communications/pbap/js/pbap.js]: P

File: apps/communications/dialer/index.html [hunk 17]
```diff
@@ -54,6 +54,9 @@
     <script defer src="/dialer/js/call_log_db.js"></script>
     <script defer src="/shared/js/multi_sim_action_button.js"></script>
 
+    <script defer src='/pbap/js/pbap.js'></script>
+    <script defer src='/pbap/js/phonebook.js'></script>
+
     <!-- for perf-measurement related utilities -->
     <script defer src='/shared/js/performance_testing_helper.js'></script>
```
Comment (pasawaytilldie23) [inline apps/communications/dialer/index.html]: M

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e52432b8f02e4247a4cd8b5a725b34e07147440d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fb42e6f6582555106b4489c0a35d086243877c0e

**Pull request merged**